### ISSUE TITLE
Manage private functions

### ISSUE THREAD

**Issue by egeerardyn** [matlab2tikz/matlab2tikz#762]
We currently have quite some private functions divided over the `test/private`, `src/private`and `test/suites/private` directories with some duplicates, near-duplicates and other nastiness.

We should come up with a strategy to manage those while reducing or removing the duplication.

I do have some ideas, but some are unfeasible in Octave, others are a bit ugly. Maybe somebody else has better ideas?

My ideas:
1. We could use `+package` directories. Those functions then get called as `pkg.function`. However, this is not supported in Octave 3.8 (I don't know about 4.0). So we would need to emulate that behavior in Octave.
2. We could just place all those functions in another directory. When executing the different public functions, we can dynamically load that path (and unload it using `onCleanup`).
3. We could only retain a single `private`directory and poke a hole to it using a wrapper function for the other directories. In those directories, the syntax would hence become more cumbersome (e.g. `[results] = m2tPrivate('functionname', arguments{:})`. This is a way that MATLAB does in the graphics toolbox.

**Comment by PeterPablo** [comment]
I like number two best. `onCleanup` already proved to be reliable and valuable. The third option makes the code illegible unnecessarily. The first option would probably require a considerable amount of work to get running.

To sum up: Two will contain very few lines of code and use an approach that we currently use successfully elsewhere.

**Comment by egeerardyn** [comment]
I think 2 is indeed best. 

Since we cannot use the `private` directory for that (due to MATLAB restrictions), I see this as an open door to also introduce some sub-directory structure to the prospective directory (we can always `genpath` everything :P), e.g. a directory for version/environment-related stuff, others for graphical stuff, ...

**Comment by PeterPablo** [comment]
yepp! sounds like a plan :+1:

**Comment by okomarov** [comment]
2nd is indeed the best. We can structure the folder as:
```
root/
|_ src/
|_ test/
|_ common/
    |_ private/
...
```
and add common only. That way, private should still be visible within function calls.

**Comment by egeerardyn** [comment]
I was thinking about a folder structure like:
```
./
|_ src/         
   |_ matlab2tikz, figure2dot, loadm2tcommon, ...
   |_ private/
      |_ m2tUpdater, ...
   |_ common/
      |_ validators/
         |_ isStream, ...
      |_ version/
         |_ isMATLAB, getEnvironment, isVersionBelow, ...
      |_ geometry/
         |_ crossLines, ...
      |_ graphics/
         |_ guitypes, ...
      |_ string/
         |_ m2tstrjoin, ...
      |_ ... (maybe other functions and folders)
|_ test/
   |_ suites/
   |_ private/
      |_ execute_tikz_stage, ...
...
```

For me the division between `./src` and `./test` is that `src` is our actual user-facing codebase and `./test` only exists to help us in developing `./src`. 

I don't think there is a need for a `private` within `common`, since our `common` folder would be a self-rolled kind of `private` that gets unloaded after our functions are ready.

**Comment by okomarov** [comment]
@egeerardyn private has the advanatge that is not callable by the user directly, so I'd favour that. It's not part of the public API and we can change it's implementation without having backward compatibility issues. 

Also, I would keep it simple for now, i.e. without categories (which are a bit reminiscent of Java projects that nest folders several times before getting to a file...). What would be the advantage?

**Comment by PeterPablo** [comment]
@okomarov, nope -- those functions would not belong to the public API, as long as the user only adds `matlab2tikz/src` to his path, as everything below `common` is added (and removed) to (from) the path during execution of m2t..

**Comment by okomarov** [comment]
@PeterPablo While we know that those do not belong to the public API, the user does not. As you say, he can add them to the path, if he finds them useful. On the ohter hand, Matlab won't allow you to add a private folder to the path, hence making it crystal clear that those are internals.

**Comment by PeterPablo** [comment]
That's a valid point: Protect the user :-)

**Comment by okomarov** [comment]
MuhaHahH, don't let them see our goodies :smiling_imp: 

Seriously speaking, it's an issue with maintenance. If we decide that some of those utilities can be of help to the user we can move them into the common folder from private, in a second step. First, we have to discuss the pro/cons, since that means once it's public we *need* to stick with that API.

**Comment by egeerardyn** [comment]
We have hundreds of functions right now, if we all dump them into a single folder, I think we might get lost. But that's the only reason I see (implementation-wise, it's just a matter of `genpath(...)` versus just `...`).

I see your point of protecting the user (or making him aware of "unsupported" features), and I agree with it. But for me one of the reasons to not use `private` within `common` would be exactly the same reason as I had for opening this issue. `private` disallows testing and code re-use (among different folders). Developers are users too: for some of our duties, we need access to those functions. Hiding them in `private` makes that a lot harder or impossible without dirty tricks (well, this topic is already a dirty trick around the lack of namespaces that work across both MATLAB and Octave).

**Comment by okomarov** [comment]
I do not see a 100 common functions though...the version functions are used by `test` and `src`, what else (I probably have a limited vision)?

I am not following:
> `private` disallows testing and code re-use  (among different folders).

**Comment by egeerardyn** [comment]
You are right there are not 100 common functions, however, we currently have a bit of an organization issue of all the functions we have. I was addressing both implicitly (my bad, as this clutters the discussion).
The 100s of functions is just the raw count, and quite a few of the functions are quite related amongst each other (e.g. Option Arrays, Version stuff, Environment stuff, ...). That's why I was proposing the folders here as well.

### `private` and testing
Functions in `private` can only be accessed from function in the parent folder. So the only way to (unit) test those is either riddle `src` with (publicly accessible) test code or to poke some dirty holes in how `private` works (e.g. `cd` into that folder, make a wrapper function, move private functions into another folder, ...). We currently don't have any unit testing, but this is something we should add in the future to avoid regressions. The problem is actually that `private` functions (and nested functions and sub-functions) are hard or impossible to access without dirty tricks. And that makes testing hard.

### Unit testing
In case you're not familiar with unit-testing, it's similar to the testing we already do but at a much finer scale. Unit tests make sure that every function performs well on some specially crafted cases (e.g. `isVersionBelow([1,0], [1,1]) == true` and quite a few other cases).

Something similar happens during the development workflow (at least in my workflow, maybe yours is different): I typically run small fragments of code from the command prompt. Private functions are not accessible from the prompt. So that means I have to either place a breakpoint at a specific place in the code and run the fragments in the debugger (which is not ideal: it takes more steps every time I debug, the workspace in the debugger is not as flexible (depending on the context it disallows to create new variables)).

### `private` and code re-use
Same problem as above: `private` functions are only accessible from parent functions. They are ment to be hidden implementation details, but cannot be shared amongst different directories without making them public. I.e. a public function in MATLAB is not always part of a public API.
